### ISSUE TITLE
postprocessing 'do-sources' with many sources fails

### ISSUE THREAD

**Issue by JoeCarlson** [intermine/intermine#861]
I've recently seen 2 issues when attempting to run the postprocessing step 'do-sources' with many sources (about a dozen).

1) db connections are opening and not closing. At a certain point, I hit my limit and cannot connect to the objectstore. (I have not yet gone to the hikari connection pooling; will this fix the problem?)

2) properties persist between sources. This may be a thing fundamental with ant and how it treats properties. If you have a source in project.xml with an optional property (such as src.data.file), the property and value persists to the next source if the next source does not overwrite the property. This makes it difficult to use optional properties.

**Comment by rnsmith** [comment]
There is definitely a problem with leaking connections during post-processing, it doesn't seem to be limited to do-sources - running enough postprocess operations of any type exhausts available connections. The same is true for integration but sources are usually run in separate JVMs by the `project_build` script.

Hikari doesn't make a difference.

The `ObjectStoreWriter` is being closed - it looks like some improvement is needed to the `close()` methods. It may be tricky as a database object may be passed into an `ObjectStore`, in which case the `ObjectStore` shouldn't really close it.

**Comment by rnsmith** [comment]
My previous comment was partially correct. The issue is actually that postprocessing steps are being run in separate ClassLoaders so are unable to share database connections. The connections don't get closed until the JVM exist and `ShutdownHook` is called so we end up creating too many connections.

I've fixed the problem for some postprocessing by not for `do-sources`, I've added a hack to force closure of database connections during `do-sources` as a workaround.

See #916.

**Comment by rnsmith** [comment]
I haven't looked at point 2 - persisting ant properties. It may be possible to fix with changes to `PostProcessTask` in `im-ant-tasks`. For the moment setting a property to an empty string in sources that don't have a value might be the best workaround.

**Comment by gcornut** [comment]
> The same is true for integration but sources are usually run in separate JVMs by the project_build script.

I can confirm that this kind of problem with database connection occurs also during data integration. I tried integrating 14 data sources at once and I encountered the error "The table intermine_metadata doesn't exist" you reported in #919.

I splited the data integration in multiple sequential ant processes for each data source and now it works fine.
